src/settings/components/RoleDetails/utils.js

```diff
@@ -38,18 +38,18 @@ export function createUserRolesRequests(previousSelectedUsers, currentSelectedUs
     }
 
     if (roleIds?.length) {
-      if (added?.includes(userId)) {
+      if (added?.includes(userId) && !roleIds.includes(roleId)) {
         roleIds.push(roleId);
       } else if (removed?.includes(userId)) {
-        roleIds.splice(roleIds.indexOf(userId), 1);
+        roleIds.splice(roleIds.indexOf(roleId), 1);
       }
       // If modified, then PUT
       if ((added.includes(userId) || removed.includes(userId)) && roleIds.length) {
         requests.push({ userId, roleIds, apiVerb: apiVerbs.PUT });
       } else if (!roleIds.length) { // If no more capabilities, DELETE
         requests.push({ userId, roleIds, apiVerb: apiVerbs.DELETE });
       }
-    } else { // if no matches, POST
+    } else if (added?.includes(userId)) { // if no matches, POST
       roleIds.push(roleId);
       requests.push({ userId, roleIds, apiVerb: apiVerbs.POST });
```

Comment:
More documentation in this section would be great. It was hard for me to follow how the manipulation of `roleIds` related to the various conditions.

Reply:
I'll include this in the follow-on story I'll create for better error handling.